torch/optim/_muon.py

```diff
@@ -0,0 +1,276 @@
+            if p.grad.is_sparse:
+                raise RuntimeError("Muon does not support sparse gradients")
+
+            params_with_grad.append(p)
+            grads.append(p.grad)
+
+            state = self.state[p]
+
+            buf = state.get("momentum_buffer")
+            if buf is None:
+                buf = torch.zeros_like(p.grad, memory_format=torch.preserve_format)
+                state["momentum_buffer"] = buf
+            muon_momentum_bufs.append(buf)
+
+        return False  # has_complex
+
+    @torch.no_grad()
+    def step(self, closure=None):
+        """Performs a single optimization step."""
+        loss = None
+        if closure is not None:
+            with torch.enable_grad():
+                loss = closure()
+
+        for group in self.param_groups:
+            lr = group["lr"]
+            weight_decay = group["weight_decay"]
+            momentum = group["momentum"]
+
+            params_with_grad: list[Tensor] = []
+            grads: list[Tensor] = []
+            muon_momentum_bufs: list[Tensor] = []
+
+            has_complex = self._init_group(
+                group,
+                params_with_grad,
+                grads,
+                muon_momentum_bufs,
+            )
+
+            muon(
+                params_with_grad,
+                grads,
+                muon_momentum_bufs,
+                lr=lr,
+                weight_decay=weight_decay,
+                momentum=momentum,
+                nesterov=group["nesterov"],
+                msign_fn=self._msign_fn,
+                msign_fn_config=self._msign_fn_config,
+                adjust_lr_fn=self._adjust_lr_fn,
+                has_complex=has_complex,
+            )
+        return loss
+
+
+def _single_tensor_muon(
+    params: list[Tensor],
+    grads: list[Tensor],
+    muon_momentum_bufs: list[Tensor],
+    *,
+    lr: float,
+    weight_decay: float,
+    momentum: float,
+    nesterov: bool,
+    msign_fn: Callable[[Tensor, BaseMsignFnConfig], Tensor],
+    msign_fn_config: BaseMsignFnConfig,
+    adjust_lr_fn: Callable[[float, torch.Size], float],
+    has_complex: bool,
+) -> None:
+    lr = _to_scalar(lr)
+    assert has_complex is False, "Complex parameters are not supported"
+
+    for i, param in enumerate(params):
+        grad = grads[i]
+        assert grad.ndim == 2, "Param gradient must be a 2D matrix"
+
+        buf = muon_momentum_bufs[i]
+        buf.mul_(momentum).add_(grad)
+        if nesterov:
+            grad = grad.add(buf, alpha=momentum)
+        else:
+            grad = buf
+
+        update = msign_fn(grad, msign_fn_config)
+        adjusted_lr = adjust_lr_fn(lr, param.shape)
+        param.mul_(1 - lr * weight_decay)
+        param.add_(update, alpha=-adjusted_lr)
+
+
+@_disable_dynamo_if_unsupported(single_tensor_fn=_single_tensor_muon)
+def muon(
+    params: list[Tensor],
+    grads: list[Tensor],
+    muon_momentum_bufs: list[Tensor],
+    foreach: Optional[bool] = None,
```

Comment:
foreach should live under the star *, and should error if passed in as true as it's not supported it. Or just don't include it.

Reply:
done!